### PR TITLE
Make mock layer compatible with DBAL 3

### LINKED ISSUE #8886: Get PHPUnit to run with DBAL 3
[Recently, a CI job was added on the 2.10.x branch to run the test suite this project after forcing the installation of `doctrine/dbal` v3](https://github.com/doctrine/orm/pull/8874).

In order not to disrupt contributors, we made sure it would not fail the build although it really does not pass at all:

https://github.com/doctrine/orm/blob/e09f126abfec0add53dc1ed8d4e3708117032fc1/.github/workflows/static-analysis.yml#L52

Right now, we get a crash without PHPUnit actually running, so we don't even know how many errors and failures there will be in the end. A good first step would be to fix that crash and get to a state where you can see the whole test suite running (with errors and failures)
If you want to help us fix it or get closer to fixing it, you can reproduce the issue locally by running (from the 2.10.x branch)

```shell
composer require doctrine/dbal ^3.0
vendor/bin/phpunit
```

When this issue is fixed, we should create a new one about making the tests pass.

### PR DESCRIPTION
**Pull request by derrabus**
😓 

Closes #8886

### CODE REVIEW

**Comment by greg0ire** [comment]
Yay PHPUnit runs! :tada: :tada: :tada: 

> There were 1285 errors

:sweat_smile:

**Comment by derrabus** [comment]
> > There were 1285 errors
> 
> 😅

If you merge my other open PRs, that number should go down to zero. 🤞🏻 😎

File: tests/Doctrine/Tests/Mocks/DriverResultMock.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function fetchOne()
+    {
+        $row = $this->fetchNumeric();
+
+        return $row ? $row[0] : false;
```
Comment (greg0ire) [inline tests/Doctrine/Tests/Mocks/DriverResultMock.php:62]: `fetchOne` is about fetching one column, so it should look a bit more like `fetchAssociative` and not stop at 0 I think.

Comment (derrabus) [reply tests/Doctrine/Tests/Mocks/DriverResultMock.php:62]: > Returns the first value of the next row of the result

I think, my implementation is correct.

Comment (greg0ire) [reply tests/Doctrine/Tests/Mocks/DriverResultMock.php:62]: Oh right I misread, it is very much correct 👍

**Review by greg0ire — APPROVED** [review]

**Pull request merged**